### ISSUE TITLE
Package Activation metrics

### ISSUE THREAD

**Issue by walles** [atom/metrics#70]
Hi!

What about collecting metrics about package activation times?

When plugins crash there is automation in place for reporting that as bugs, but when plugins are slow to activate, that's really un-obvious (for package authors).

If package activation time stats could be published it would help package authors know if their packages are slow to activate.

In my case, I just discovered `timecop`, uninstalled `git-plus` and a number of other packages, and got my startup time improved by a lot. However, if these stats were easily available, maybe the authors of those packages could have fixed their packages instead, thereby improving *everybody*'s startup time (and enabled me to keep using their packages).

**Comment by walles** [comment]
Pull request: https://github.com/atom/metrics/pull/71

**Comment by thomasjo** [comment]
I'm on 👎 on this since the intended audience for the collected metrics is GitHub, and data is collected in order to gather useful information that might help improve the core experience. The intent is not to collect wide swath of metadata for the purpose of publication.

Package authors who actually care about their users can easily benchmark their packages using e.g. timecop. By publishing such statistics we risk, in some sense, of ending up with a "wall of shame".

If a package author doesn't care their package is slow, or does not have the resources to fix the problem, submit a PR (or fork the package).

/cc @atom/feedback

**Comment by walles** [comment]
Depends on whether you want to optimize the Atom *core* experience or the *Atom* experience.

And you *are* automatically informing plugin authors about crashes in their plugins, which indicates you actually do care about the full Atom experience rather than just the core one.

I'd argue that neither plugin authors nor plugin users actually know about the startup time footprints, so they aren't even *able* to care, and publicizing package startup time information would be really beneficial to everybody.

But all of this is of course up to you to decide, I'm fine with a decision either way.

**Comment by thomasjo** [comment]
First of all, we are not _automatically_ informing package authors about runtime bugs. Rather end-users are presented with an option to open a (partially filled in) issue on the package repo.

There's a fundamental difference between a package having bugs (or crashing), and having poor (start-up) performance. Crashes/bugs can to a large extent be affected by other packages, and the number of package permutations is extremely high. That makes it very difficult for a package author to make their product bug free, etc.

On the other hand, checking the activation/start-up performance is trivial in comparison. Package activation is (when done appropriately) unaffected by other packages. Unless a package absolutely must activate when Atom loads, it should use deferred activation via supported mechanisms. All of this can be easily tested and benchmarked by package authors _prior_ to releasing code. There might be edge cases here, but nothing that (in my opinion) requires collecting metrics at this scale.

**Comment by walles** [comment]
I still think:
1. TimeCop is great!
2. It's not necessarily obvious to plugin authors that they could / should run it to check for startup time.

Also, if one plugin is very slow that's easy enough for a plugin author to find and fix, but if lots of plugins are kind of slow that can still affects end users a lot without it being obvious at all to plugin writers, even the ones who actually *do* profile their plugins.

**Comment by Ben3eeE** [comment]
> It's not necessarily obvious to plugin authors that they could / should run it to check for startup time.

What we can do and would accept a PR for is to better document timecop, deferred activation and other start up performance related issues and fixes in the Atom flight manual to make it more obvious to package authors. It is currently mostly covered in the [debugging section](http://flight-manual.atom.io/hacking-atom/sections/debugging/) which might not be the first thing an aspiring package author checks before publishing. How can we make it more obvious?

**Comment by thomasjo** [comment]
If you think it's not obvious to package authors that they can/should benchmark their packages, using for example timecop, then that is solved by _educating_ them. It should not be solved by collecting relevant metrics, and then publicly "shaming" slow packages (and a package might have a perfectly valid reason for being slow, which is well understood by its users).

The proposed solution is (collecting metrics and publishing it) is masking the real the problem, which is that we probably need better/more documentation around "best" practice package development, and so on.

**Comment by 50Wliu** [comment]
I agree with all the points @thomasjo has presented here, and am similarly :-1: on including this in metrics.  If it is non-obvious that Timecop displays package load times, we should make it more obvious as @Ben3eeE said.

**Comment by damieng** [comment]
We could take a leaf out of the Windows playbook and put up a warning if Atom is slow to start and offer to pop up Timecop.

**Comment by thomasjo** [comment]
> We could take a leaf out of the Windows playbook and put up a warning if Atom is slow to start and offer to pop up Timecop.

Unless this only enabled in dev mode, we would need to introduce a permanent way to disable this globally (or possibly whitelisting certain packages), since some packages might "have to" be slow to activate. And that might be fine for users of said hypothetical package. We shouldn't annoy end-users because a package they are using is slow. That just adds injury to insult, right?

At the end of the day, this boils down to educating and assisting package developers.

**Comment by damieng** [comment]
I was thinking we could just put something in the status bar that says "Atom took 2.3 seconds to start. Show package start-up performance." with the second part being a link to timecop.

That way it's non-invasive.

"We shouldn't annoy end-users because a package they are using is slow. That just adds injury to insult, right?"

I'd say they are already annoyed but put the blame squarely at Atom's feet.

**Comment by thomasjo** [comment]
My point is that someone might install package `xyz` well aware of it being slow, accepting that fact, and placing zero blame on Atom.

I think your suggestion is reasonable, albeit somewhat orthogonal to the intent of this issue though (which is focused on package authors, not the end-user). Since status-bar real estate is a very limited resource, we'd have to limit ourselves to an octicon perhaps?

**Comment by damieng** [comment]
I think we can take over the main status-bar area at Atom launch time then let it get wiped out by whatever else uses it when the user starts using Atom.

That way it's both clear and it doesn't bug users by sitting around all day as a bright icon they might have already 'accepted'

**Comment by walles** [comment]
One way to focus on package developers would be to:
1. If Atom is started in dev mode...
2. ... then for each package loaded from the dev directory...
3. ... warn if the activation time is "too long".

This message would reach only people who can actually *do* something about the package's performance.

How the notification would be done I'm not sure of, but it could be louder than something intended for all users.
